### PR TITLE
systemtest/loadgen: don't pre-compress batches

### PR DESCRIPTION
**Pull request by axw**
## Motivation/summary

Compress batches while sending, rather than ahead of time. This will enable us to transform events (https://github.com/elastic/apm-server/issues/10445 and https://github.com/elastic/apm-server/issues/10446) and may also simplify state-state event generation (https://github.com/elastic/apm-server/pull/10468#pullrequestreview-1348434618).

This changes does add significant runtime overhead to the load generator, but:
 - we mitigate that somewhat by using `zlib.BestSpeed`, and using the more optimised `klauspost/compress` package
 - `apmbench` is ultimately limited by APM Server's throughput, and the difference doesn't impact real world benchmarks. I ran `testing/benchmark` twice, once on main and once with `apmbench` built off this branch, with `BENCHMARK_COUNT=6 BENCHMARK_RUN=BenchmarkAgentAll`. There were no significant differences:

```
            │ /tmp/old.txt │        /tmp/new.txt         │
            │    sec/op    │   sec/op    vs base         │
AgentAll-64     12.39 ± 1%   12.42 ± 2%  ~ (p=0.485 n=6)

            │    /tmp/old.txt     │              /tmp/new.txt              │
            │ error_responses/sec │ error_responses/sec  vs base           │
AgentAll-64            0.000 ± 0%            0.000 ± 0%  ~ (p=1.000 n=6) ¹
¹ all samples are equal

            │ /tmp/old.txt │        /tmp/new.txt         │
            │  errors/sec  │ errors/sec  vs base         │
AgentAll-64     29.05 ± 1%   28.99 ± 2%  ~ (p=0.513 n=6)

            │ /tmp/old.txt │         /tmp/new.txt         │
            │  events/sec  │ events/sec   vs base         │
AgentAll-64    1.463k ± 2%   1.458k ± 2%  ~ (p=0.372 n=6)

            │ /tmp/old.txt │        /tmp/new.txt         │
            │  gc_cycles   │ gc_cycles   vs base         │
AgentAll-64     57.00 ± 7%   57.00 ± 5%  ~ (p=0.818 n=6)

            │  /tmp/old.txt  │          /tmp/new.txt           │
            │ max_goroutines │ max_goroutines  vs base         │
AgentAll-64       57.00 ± 2%       58.00 ± 9%  ~ (p=0.431 n=6)

            │  /tmp/old.txt  │          /tmp/new.txt           │
            │ max_heap_alloc │ max_heap_alloc  vs base         │
AgentAll-64      311.2M ± 6%      300.8M ± 4%  ~ (p=0.589 n=6)

            │   /tmp/old.txt   │           /tmp/new.txt            │
            │ max_heap_objects │ max_heap_objects  vs base         │
AgentAll-64       2.512M ± 11%       2.340M ± 10%  ~ (p=0.310 n=6)

            │ /tmp/old.txt │         /tmp/new.txt         │
            │   max_rss    │   max_rss    vs base         │
AgentAll-64    363.8M ± 4%   368.4M ± 4%  ~ (p=0.699 n=6)

            │      /tmp/old.txt       │               /tmp/new.txt               │
            │ mean_available_indexers │ mean_available_indexers  vs base         │
AgentAll-64                9.139 ± 1%                9.136 ± 1%  ~ (p=0.675 n=6)

            │ /tmp/old.txt │         /tmp/new.txt         │
            │ metrics/sec  │ metrics/sec  vs base         │
AgentAll-64     230.6 ± 2%    229.1 ± 2%  ~ (p=0.394 n=6)

            │ /tmp/old.txt │        /tmp/new.txt         │
            │  spans/sec   │ spans/sec   vs base         │
AgentAll-64     871.7 ± 1%   869.8 ± 2%  ~ (p=0.485 n=6)

            │ /tmp/old.txt │        /tmp/new.txt         │
            │   txs/sec    │  txs/sec    vs base         │
AgentAll-64     331.4 ± 1%   330.8 ± 2%  ~ (p=0.513 n=6)

            │ /tmp/old.txt │         /tmp/new.txt          │
            │     B/op     │     B/op      vs base         │
AgentAll-64   554.4Mi ± 0%   553.8Mi ± 0%  ~ (p=0.485 n=6)

            │ /tmp/old.txt │           /tmp/new.txt            │
            │  allocs/op   │  allocs/op   vs base              │
AgentAll-64    7.986M ± 0%   7.994M ± 0%  +0.10% (p=0.004 n=6)
```

## Checklist

~- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)~
~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~
~- [ ] Documentation has been updated~

### CODE REVIEW

File: systemtest/loadgen/eventhandler/handler.go [hunk 1]
```diff
@@ -20,120 +20,106 @@ package eventhandler
...
-				})
+			// Copy the line, as it will be overwritten by the next scan.
+			linecopy := make([]byte, len(line))
+			copy(linecopy, line)
+			line = linecopy
+
+			if isMeta := bytes.HasPrefix(line, metaHeader); isMeta {
+				h.batches = append(h.batches, batch{})
+				current = &h.batches[len(h.batches)-1]
+				current.metadata = linecopy
```
Comment (marclop) [inline systemtest/loadgen/eventhandler/handler.go:118]: Maybe?
```suggestion
				h.batches = append(h.batches, batch{metadata: linecopy})
				current = &h.batches[len(h.batches)-1]
```

File: systemtest/loadgen/eventhandler/handler.go [hunk 2]
```diff
@@ -20,120 +20,106 @@ package eventhandler
...
-					return nil, err
-				}
-				h.batches = append(h.batches, batch{
-					bytes: writer.Bytes(),
-					items: scanned,
-				})
+			// Copy the line, as it will be overwritten by the next scan.
+			linecopy := make([]byte, len(line))
+			copy(linecopy, line)
+			line = linecopy
```
Comment (marclop) [inline systemtest/loadgen/eventhandler/handler.go:113]: Why is this assignment necessary? It seems that we're copying `line` to `linecopy`, but `copy()` doesn't zero the `line` variable, right?

**Review by marclop — APPROVED** [review]
Looks great. Thank you for the refactor

File: systemtest/loadgen/eventhandler/handler.go [hunk 3]
```diff
@@ -20,120 +20,106 @@ package eventhandler
...
+
 	h := Handler{
-		transport:  t,
-		limiter:    l,
-		readerPool: newByteReaderPool(),
+		config: config,
+		writerPool: sync.Pool{
+			New: func() any {
+				pw := &pooledWriter{}
+				zw, err := zlib.NewWriterLevel(&pw.buf, zlib.BestSpeed)
```
[the remark this answers is not in the document]
Comment (axw) [reply systemtest/loadgen/eventhandler/handler.go:78]: I don't know what the reality is, but the spec (https://github.com/elastic/apm/blob/main/specs/agents/transport.md#compression) says:

> Agents MUST compress the HTTP payload and SHOULD optimize for speed over compactness (typically known as the "best speed" level).

So I think we should just use that without any config, and consider making it optional if we consider changing the recommendation for agents.

File: systemtest/loadgen/eventhandler/handler.go (hunk 2 above)
Comment (axw) [reply systemtest/loadgen/eventhandler/handler.go:113]: It's not, I had intended to use `line` everywhere below, but I didn't :smile:
I'll remove this.

File: systemtest/loadgen/eventhandler/handler.go [hunk 4]
```diff
@@ -20,120 +20,105 @@ package eventhandler
...
 
-			if isMeta := bytes.HasPrefix(line, metaHeader); !isMeta {
-				writer.WriteLine(line)
-				scanned++
-				continue
-			}
+			// Copy the line, as it will be overwritten by the next scan.
+			linecopy := make([]byte, len(line))
+			copy(linecopy, line)
+			line = linecopy
```
Comment (axw) [inline systemtest/loadgen/eventhandler/handler.go:113]:
```suggestion
```

**Pull request merged**